### PR TITLE
Added pinned comment moderation

### PR DESCRIPTION
**Pull request by JohnONolan**
## Summary
Adds support for pinning member comments — both on the front-end (`comments-ui`) and in admin (`apps/posts`) — for users with moderation access. Includes the migration, schema, model, service, controller, endpoints, mapper, and matching UI.

Refined the moderation context menu design (icons for actions) so the comments-ui menu reads consistently with the admin moderation flow.

All behaviour is gated behind a new private labs flag, `commentsPinning`.

## Behind a labs flag
- Backend: `comments-service.moderateComment` rejects `pinned`-field writes with `MethodNotAllowedError` when the flag is off; `pinnedFirst` ordering only applies when on, so existing `pinned_at` values don't leak ordering changes if the flag is later disabled.
- comments-ui: `PinnedLabel` returns null and the admin context menu hides Pin/Unpin items when the flag is off.
- apps/posts: badge + dropdown items are gated via a shared `useCommentsPinningEnabled()` hook.

## Carved-out work (already merged)
The original PR bundled three things; the two non-pinning pieces landed separately so this PR is just pinning + labs gating:
- **Default `comments_enabled` → `"all"`** + admin sidebar visibility — #27913 (merged)
- **`commentsPinning` private labs flag registration** — #27914 (merged)

## Front end
CleanShot 2026-05-02 at 16 59 39@2x

## Admin
CleanShot 2026-05-02 at 17 17 27@2x

### CODE REVIEW

File: ghost/core/core/server/api/endpoints/utils/serializers/output/mappers/comments.js [hunk 1]
```diff
@@ -78,7 +88,15 @@ const commentMapper = (model, frame) => {
...
-    const response = _.pick(jsonModel, commentFields);
+    const fields = getRequestedFields(frame);
+    const response = _.pick(jsonModel, fields ? commentFields.filter(field => fields.has(field)) : commentFields);
+
+    if (!fields || fields.has('pinned')) {
+        const canShowPinned = !jsonModel.parent_id && Boolean(jsonModel.pinned_at);
+        response.pinned = isPublicRequest
+            ? canShowPinned && jsonModel.status === 'published'
+            : canShowPinned;
+    }
```
[the remark this answers is not in the document]
Comment (JohnONolan) [reply ghost/core/core/server/api/endpoints/utils/serializers/output/mappers/comments.js:104]: Guarded redacted html writes behind requested fields and added public fields coverage for hidden and deleted comment placeholders.

File: ghost/core/core/server/services/comments/comments-service.js [hunk 2]
```diff
@@ -216,11 +219,55 @@ class CommentsService {
...
+                }
+
+                if (existingComment.get('status') === 'deleted' || data.status === 'deleted') {
+                    throw new errors.BadRequestError({
+                        message: tpl(messages.cannotPinDeletedComment)
+                    });
+                }
+            }
+
+            editData.pinned_at = data.pinned ? new Date() : null;
```
[the remark this answers is not in the document]
Comment (JohnONolan) [reply ghost/core/core/server/services/comments/comments-service.js]: Preserved the existing pinned_at timestamp on repeat pin requests and added admin API coverage for idempotent pinning.

File: ghost/core/core/server/services/comments/comments-service.js [hunk 3]
```diff
@@ -12,9 +12,26 @@ const messages = {
...
+    }
+
+    const statusClause = options.isAdmin ? '' : ' AND comments.status = \'published\'';
+    const pinnedSelect = `CASE WHEN comments.parent_id IS NULL AND comments.pinned_at IS NOT NULL${statusClause} THEN 1 ELSE 0 END AS pinned`;
+
+    return {
+        ...options,
+        selectRaw: [options.selectRaw, pinnedSelect].filter(Boolean).join(', ')
+    };
+}
```
[the remark this answers is not in the document]
Comment (JohnONolan) [reply ghost/core/core/server/services/comments/comments-service.js]: Applied the pinned-field select shim to single-comment reads and added members API coverage for fields=id,pinned on a single comment.

**Comment by cathysarisky** [comment]
I'm curious - what's the rationale behind having comments on by default? Probably half the folks who do not want comments because they're a moderation headache and potential spam source.  Some of the other half are very interested in comments as an engagement mechanism, so improving comments is awesome -- I'm just not sure that making comments opt-out is desirable.  It's a lot easier to tell someone how to turn on commenting than to tell someone how to clear out a bunch of spammy comments that they didn't know were enabled.

**Comment by ErisDS** [comment]
Updated state of the task list.

**Extracted to separate PRs (done):**
- [x] Default-flip + sidebar visibility — #27913 (merged)
- [x] `commentsPinning` private labs flag — #27914 (merged)

**Block before merge:**
- [x] Gate the pinning UI and service on `labs.commentsPinning`.
- [x] Fix migration `down()` symmetry: the composite `[post_id, parent_id, pinned_at]` index is the only one covering the `post_id` FK, so `down()` adds a single-column `[post_id]` index first to keep the FK covered when dropping the composite (MySQL otherwise rejects with "needed in a foreign key constraint"). Verified locally against MySQL 8.
- [x] Restore the integration test at `test/integration/services/comments-service.test.js` (lost when the default-flip was pulled out).

**Should fix:**
- [x] Extract `PinnedLabel` to its own file (`apps/comments-ui/src/components/content/pinned-label.tsx`) — mirrors how threading split out `threaded-replies.tsx` / `reply-tree.tsx` / `focused-thread.tsx`.
- [x] Single `useCommentsPinningEnabled()` hook in `apps/posts` — collapses three `useBrowseConfig()` reads (`comment-menu.tsx`, `comments-list.tsx`, `comment-thread-list.tsx`) into one shared hook. Reduces GA cleanup to a single file.
- [x] Add an e2e API test asserting `MethodNotAllowedError` when `commentsPinning` is disabled. Also hoisted `mockLabsEnabled('commentsPinning')` into the file-level `beforeEach` since `mockManager.restore()` runs at `after()` not `afterEach`, so without explicit enable per test the disable-case would leak into siblings.
- [x] Add a unit test asserting `PinnedLabel` returns null when the labs flag is off (new file `pinned-label.test.jsx`; the context menu's labs-off case is also covered in `comment-context-menu.test.jsx`).

**Verify / decide:**
- [x] ~~Smoke-test the new `useGlobalData()` call sites~~ — done by it breaking and getting fixed. `useGlobalData()` reads from `PostAnalyticsProvider`, which only mounts on the per-post analytics route (requires `postId` from `useParams`). The site-wide `/ghost#/comments` view crashed on mount because the provider wasn't there. All three call sites now use `useBrowseConfig()` via the shared `useCommentsPinningEnabled()` hook.
- [x] Confirm intent for read-path gating: `commentMapper` still exposes the `pinned` field when the flag is off; only `pinnedFirst` ordering is gated. Matches the threading PR pattern. **Decision: leave as-is.**

**Nice-to-have:**
- [x] Unify the three gating idioms — comments-ui now uses `labs?.commentsPinning === true` / `!== true` (matching `useCommentsPinningEnabled()` in apps/posts and the established Ghost pattern in `members-filters.tsx` / `import-members-modal.tsx`). Server stays on the canonical `this.labs?.isSet('commentsPinning')` API.
- [x] Reuse `handleCacheHeaders` in `adminEdit` — moved the X-Cache-Invalidate helper out of the comments endpoint and onto `CommentsController.setCacheInvalidationHeaders`. Both the endpoint's `add` path and `adminEdit` now call the same method, so the two cache invalidations cannot drift.

**Housekeeping:**
- [x] Moved the migration from `6.39/` to `6.40/` after v6.39.0 shipped mid-PR — once 6.39.0 is published, any new migration in `6.39/` would be skipped on already-migrated installs. `pnpm migrate:create` repositioned it and bumped `ghost/core` + `ghost/admin` to `6.40.0-rc.0`.
